packages/@react-stately/toast/src/useToastState.ts

```diff
@@ -159,30 +155,27 @@ export class ToastQueue<T> {
     if (index >= 0) {
       this.queue[index].onClose?.();
       this.queue.splice(index, 1);
+      let closedToast = this.visibleToasts.find(t => t.key === key);
+      if (closedToast) {
+        closedToast.animation = 'exiting';
+      }
     }
 
-    this.updateVisibleToasts(index);
+    this.updateVisibleToasts('close');
   }
 
   /** Removes a toast from the visible toasts after an exiting animation. */
   remove(key: string) {
-    let index = this.queue.findIndex(t => t.key === key);
-    this.visibleToasts = this.visibleToasts.filter(t => t.key !== key);
-    this.updateVisibleToasts(index);
+    this.queue = this.queue.filter(t => t.key !== key);
+    this.updateVisibleToasts('remove');
   }
 
-  private updateVisibleToasts(oldIndex = -1) {
+  private updateVisibleToasts(action: 'add' | 'close' | 'remove') {
     let toasts = this.queue.slice(0, this.maxVisibleToasts);
-    if (this.hasExitAnimation) {
-      let prevToasts: QueuedToast<T>[] = this.visibleToasts
-          .filter(t => !toasts.some(t2 => t.key === t2.key))
-          .map(t => ({...t, animation: 'exiting'}));
 
-      if (oldIndex !== -1) {
-        toasts.splice(oldIndex, 0, prevToasts?.[0]);
-      }
-
-      this.visibleToasts = toasts;
+    if (action === 'close' && this.hasExitAnimation) {
+      // Cause a rerender to happen for exit animation
+      this.visibleToasts = this.visibleToasts.map(t => t);
```

Comment:
One change in logic here is that previously, if the number of visible toasts exceeded the `maxVisibleToasts` setting, the toasts that got pushed off the end would animate out. Now they won't. Only toasts that are explicitly closed by clicking the X button will animate. Is that expected?

Reply:
Oh didn't realize that was a case here. Will revisit how we apply the `.animation` then and see if we can move it back into here somehow